source/common/protobuf/utility.h

```diff
@@ -429,9 +429,27 @@ class MessageUtil {
    * values, e.g. an int32 set to 0 or a bool set to false.
    * @return std::string of formatted JSON object.
    */
-  static std::string getJsonStringFromMessage(const Protobuf::Message& message,
-                                              bool pretty_print = false,
-                                              bool always_print_primitive_fields = false);
+  static ProtobufUtil::StatusOr<std::string>
+  getJsonStringFromMessage(const Protobuf::Message& message, bool pretty_print = false,
+                           bool always_print_primitive_fields = false);
+
+  /**
+   * Extract JSON as string from a google.protobuf.Message, crashing if the conversion to JSON
```

Comment:
This contract is not strong enough IMHO. I think we should not have any `OrDie` unless the caller can safely predict that the JSON conversion will succeed. So far, we've only identified a single case, the "unknown type in Any", where this is not guaranteed to succeed. Can we document this as a precondition and guarantee conversion? Otherwise I would vote to eliminate `OrDie` and change all sites to something safe.

Reply:
Just for added info, I have only seen OOM situations from deep recursion in fuzz bugs:
https://github.com/envoyproxy/envoy/pull/9422#discussion_r360448124

Failure occurs in the docs  https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.util.json_util#BinaryToJsonStream.details if
```
1. TypeResolver fails to resolve a type.
2. input is not valid protobuf wire format, or conflicts with the type
   information returned by TypeResolver.
```

Reply:
There are a lot of call sites where we construct a proto and then immediately turn it into JSON. It didn't seem necessary to require extra ceremony there.

Reply:
Sorry, sent prematurely. I'm not a fan of `OrDie` in general; I mostly didn't want to create a giant PR that changes all the call sites at once. This at least makes it clear that there is some danger. Here's a semi-crazy idea: we can inspect protobuf types at runtime to determine which of them contain `Any` messages in them. Instead of doing that when running Envoy, add a build step that generates an "approved" list of protos by doing the traversal, and then writes out overloads of some `getSafeJsonStringFromMessage` function that can just return the JSON string.

Reply:
I think we could do something like that, but my original comment wasn't actually asking for a big change here, just that we make the contract clearer in the documentation comments. Basically, state that we do in fact always succeed if the types are known, so then a caller can know it's safe to use `OrDie` in the situations similar to what you describe, i.e. when we have a known proto and we're just dumping it.

Reply:
Thanks, I misunderstood. I've updated the comment.